File: src/board/export.ts
import { exportMapper } from '../tracking/frameMapper';
import type { BoardAction } from '../types';
import { paintActions } from './board';

export type ExportMode = 'ink' | 'composite';

export interface ExportRequest {
  actions: readonly BoardAction[];
  videoWidth: number;
  videoHeight: number;
  mirrored: boolean;
  mode: ExportMode;
  /** Required for `composite`, ignored otherwise. */
  video?: HTMLVideoElement | null;
}

/**
 * Renders at the camera's native resolution rather than screenshotting the
 * stage, so the export is independent of window size and of the cropping that
 * object-fit: cover applies. `ink` keeps the alpha channel — the whiteboard
 * really is transparent.
 */
export async function exportPng(request: ExportRequest): Promise<Blob> {
  const { actions, videoWidth, videoHeight, mirrored, mode, video } = request;

  const width = Math.max(videoWidth, 1);
  const height = Math.max(videoHeight, 1);

  const canvas = document.createElement('canvas');
  canvas.width = width;
  canvas.height = height;

  const ctx = canvas.getContext('2d');
  if (!ctx) throw new Error('Canvas 2D is unavailable in this browser.');

  if (mode === 'composite' && video && video.readyState >= HTMLMediaElement.HAVE_CURRENT_DATA) {
    ctx.save();
    if (mirrored) {
      ctx.translate(width, 0);
      ctx.scale(-1, 1);
    }
    ctx.drawImage(video, 0, 0, width, height);
    ctx.restore();
  }

  const mapper = exportMapper(width, height, mirrored);

  if (mode === 'composite') {
    // Erase actions must not cut a hole through the video frame underneath, so
    // ink is composed on its own layer first and then laid on top.
    const inkLayer = document.createElement('canvas');
    inkLayer.width = width;
    inkLayer.height = height;
    const inkCtx = inkLayer.getContext('2d');
    if (!inkCtx) throw new Error('Canvas 2D is unavailable in this browser.');
    paintActions(inkCtx, actions, mapper, width, height);
    ctx.drawImage(inkLayer, 0, 0);
  } else {
    paintActions(ctx, actions, mapper, width, height);
  }

  return new Promise<Blob>((resolve, reject) => {
    canvas.toBlob(
      (blob) => (blob ? resolve(blob) : reject(new Error('The image could not be encoded.'))),
      'image/png',
    );
  });
}

export function downloadBlob(blob: Blob, filename: string): void {
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = filename;
  document.body.append(link);
  link.click();
  link.remove();
  // Give the browser a beat to start the download before revoking.
  setTimeout(() => URL.revokeObjectURL(url), 10_000);
}

export function timestampedName(mode: ExportMode): string {
  const now = new Date();
  const pad = (n: number): string => String(n).padStart(2, '0');
  const stamp =
    `${now.getFullYear()}${pad(now.getMonth() + 1)}${pad(now.getDate())}` +
    `-${pad(now.getHours())}${pad(now.getMinutes())}${pad(now.getSeconds())}`;
  return `drawone-${mode}-${stamp}.png`;
}

File: drawone/board/export.py
"""PNG export.

Renders at the camera's native resolution rather than screenshotting the window,
so the export is independent of window size and of the cropping that cover-fit
applies. `ink` keeps the alpha channel — the whiteboard really is transparent.
"""

from __future__ import annotations

import datetime as _datetime
from pathlib import Path
from typing import Literal, Optional, Sequence

import cv2
import numpy as np

from ..render.canvas import Layer, composite_over, to_straight_bgra
from ..tracking.frame_mapper import export_mapper
from ..types import BoardAction
from .board import paint_actions

ExportMode = Literal["ink", "composite"]


def render_export(
    actions: Sequence[BoardAction],
    video_width: int,
    video_height: int,
    mirrored: bool,
    mode: ExportMode,
    frame: Optional[np.ndarray] = None,
) -> np.ndarray:
    """The finished image: BGRA for `ink`, BGR for `composite`."""
    width = max(int(video_width), 1)
    height = max(int(video_height), 1)

    layer = Layer(width, height)
    paint_actions(layer, actions, export_mapper(width, height, mirrored))

    if mode != "composite":
        return to_straight_bgra(layer)

    if frame is None:
        base = np.zeros((height, width, 3), np.uint8)
    else:
        base = frame
        if base.shape[0] != height or base.shape[1] != width:
            base = cv2.resize(base, (width, height), interpolation=cv2.INTER_AREA)
        if mirrored:
            base = cv2.flip(base, 1)
        base = np.ascontiguousarray(base)

    # Ink lives on its own layer, so an erase cuts holes in the drawing rather
    # than through the camera frame underneath.
    composite_over(base, layer)
    return base


def export_png(
    path: Path,
    actions: Sequence[BoardAction],
    video_width: int,
    video_height: int,
    mirrored: bool,
    mode: ExportMode,
    frame: Optional[np.ndarray] = None,
) -> Path:
    image = render_export(actions, video_width, video_height, mirrored, mode, frame)
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    if not cv2.imwrite(str(path), image):
        raise OSError("The image could not be encoded: {}".format(path))
    return path


def timestamped_name(mode: ExportMode) -> str:
    stamp = _datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
    return "drawone-{}-{}.png".format(mode, stamp)
